internal/controller/accessrequest_controller.go

```diff
@@ -33,30 +38,241 @@ type AccessRequestReconciler struct {
+	if err != nil {
+		logger.Error(err, "handlePermission error")
+		return ctrl.Result{}, fmt.Errorf("error handling permission: %w", err)
+	}
+
+	return buildResult(status, &ar), nil
+}
+
+// buildResult will verify the given status and determine when this access
+// request should be requeued.
+func buildResult(status api.Status, ar *api.AccessRequest) ctrl.Result {
+	result := ctrl.Result{}
+	switch status {
+	case api.RequestedStatus:
+		result.Requeue = true
+		// TODO add a controller configuration
+		result.RequeueAfter = time.Minute * 3
+	case api.GrantedStatus:
+		result.Requeue = true
+		result.RequeueAfter = ar.Spec.Duration.Duration
+	}
+	return result
+}
+
+func (r *AccessRequestReconciler) handlePermission(ctx context.Context, ar *api.AccessRequest) (api.Status, error) {
 
-	// TODO(user): your logic here
+	resp, err := r.Allowed(ctx, ar)
+	if err != nil {
+		return "", fmt.Errorf("error verifying if subject is allowed: %w", err)
+	}
+	if !resp.Allowed {
+		err = r.updateStatusWithRetry(ctx, ar, api.DeniedStatus, resp.Message)
+		if err != nil {
+			return "", fmt.Errorf("error updating access request status to denied: %w", err)
+		}
+		return api.DeniedStatus, nil
+	}
+
+	err = r.updateAppProjectRBAC(ctx, ar)
+	if err != nil {
+		return "", fmt.Errorf("error updating AppProject RBAC: %w", err)
+	}
+	err = r.updateStatusWithRetry(ctx, ar, api.GrantedStatus, resp.Message)
+	if err != nil {
+		return "", fmt.Errorf("error updating access request status to granted: %w", err)
+	}
+	return api.GrantedStatus, nil
+}
+
+func (r *AccessRequestReconciler) updateStatusWithRetry(ctx context.Context, ar *api.AccessRequest, status api.Status, details string) error {
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		err := r.Get(ctx, client.ObjectKeyFromObject(ar), ar)
+		if err != nil {
+			return err
+		}
+		ar.UpdateStatus(status, details)
+		return r.Status().Update(ctx, ar)
+	})
+}
+
+// TODO
+func (r *AccessRequestReconciler) updateAppProjectRBAC(ctx context.Context, ar *api.AccessRequest) error {
+	// 1. retrieve the Application
+	// 2. verify if CR is approved
+	// 3. retrieve the AppProject
+	// 4. assign user in the desired role in the AppProject
+	return nil
+}
+
+type AllowedResponse struct {
+	Allowed bool
+	Message string
+}
+
+// TODO
+func (r *AccessRequestReconciler) Allowed(ctx context.Context, ar *api.AccessRequest) (AllowedResponse, error) {
+	return AllowedResponse{}, nil
+}
+
+// handleAccessExpired will verify if the given AccessRequest is expired. If
+// so, it will remove the Argo CD access for the subject and update the
+// AccessRequest status field.
+func (r *AccessRequestReconciler) handleAccessExpired(ctx context.Context, ar *api.AccessRequest) (bool, error) {
+	expired := false
+	if ar.Status.ExpiresAt != nil &&
+		ar.Status.ExpiresAt.Time.After(time.Now()) {
+
+		err := r.removeArgoCDAccess(ar)
+		if err != nil {
+			return expired, fmt.Errorf("error removing access for expired request: %w", err)
```

Comment:
I think `expired = true` should be set at the beginning of the if condition. even if err is not nil, it might be confusing to have expired set to false.

Reply:
I think both cases can cause confusion. I opted for rely on the golang zero value initialization for boolean pattern. When a function returns 2 params with `error`, usually the general practice is "ignore all the rest" and handle the error.  I updated the godoc with the proper explanation.

Reply:
refactored according to our call